### PR TITLE
Implement copylat for latency measurement

### PR DESCRIPTION
**Pull request by pakmarkthub**
This PR:
- adds copylat for latency measurement.
- consolidates definitions in common.hpp.

### CODE REVIEW

File: tests/copylat.cpp [hunk 1]
```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) 2014, NVIDIA CORPORATION. All rights reserved.
```
Comment (drossetti) [inline tests/copylat.cpp]: 2020

File: tests/copylat.cpp [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+#if defined(GDRAPI_X86)
+#define FENCE() asm volatile("mfence":::"memory")
+#elif defined(GDRAPI_POWER)
+#define FENCE() asm volatile("sync":::"memory")
+#else
+#define FENCE() asm volatile("":::"memory")
+#endif
+
+#define OUT cout
```
Comment (drossetti) [inline tests/copylat.cpp]: can you remove this macro and its uses?

**Comment by drossetti** [comment]
@pakmarkthub I don't believe copylat addresses #26, but it is highly needed anyways

**Comment by pakmarkthub** [comment]
@drossetti Fixed the code.

**Comment by pakmarkthub** [comment]
> 
> 
> I had that other comment about adding a paragraph to the readme where you show the output?

Sorry, I missed the comment.

Updated README.md.

File: README.md [hunk 3]
```diff
@@ -123,6 +123,80 @@ BAR1 read BW: 1.50172MB/s
...
+map_d_ptr: 0x7f6ae5000000
+info.va: 7f6aca800000
+info.mapped_size: 16777216
+info.page_size: 65536
+info.mapped: 1
+info.wc_mapping: 1
+page offset: 0
+user-space pointer: 0x7f6ae5000000
+
+gdrcopy_H2D num iters for each size: 10000
```
Comment (drossetti) [inline README.md]: Can we use the name of the APIs rather D2H etc?

**Comment by pakmarkthub** [comment]
Renamed the output.

@drossetti Please take a look.

File: tests/copylat.cpp [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+using namespace gdrcopy::test;
+
+#if defined(GDRAPI_X86)
+#define FENCE() asm volatile("mfence":::"memory")
+#elif defined(GDRAPI_POWER)
+#define FENCE() asm volatile("sync":::"memory")
+#else
+#define FENCE() asm volatile("":::"memory")
+#endif
```
Comment (drossetti) [inline tests/copylat.cpp]: wondering whether these macros as well as MYCLOCK can be moved to common.hpp

**Comment by pakmarkthub** [comment]
@drossetti 

I use this chance to consolidate definitions from other test files into common.hpp.

Requested re-review.

File: tests/copylat.cpp [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+        cout << "gdr_copy_to_mapping num iters for each size: " << num_write_iters << endl;
+        printf("Test \t\t\t Size(B) \t Avg.Time(us)\n");
+        copy_size = 1;
+        while (copy_size <= size) {
+            int iter = 0;
+            uint8_t dat;
+            clock_gettime(MYCLOCK, &beg);
+            for (iter = 0; iter < num_write_iters; ++iter) {
+                gdr_copy_to_mapping(mh, buf_ptr, init_buf, copy_size);
+                // ensure that the data reaches the GPU's point of consistency
```
Comment (drossetti) [inline tests/copylat.cpp]: this is beyond the aim of this change, as the 3rd party access to device memory is not fully defined at the moment.

File: tests/copylat.cpp [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        } END_CHECK;
+
+        cout << endl;
+    }
+
+    cout << endl;
+
+    gdr_t g = gdr_open();
+    ASSERT_NEQ(g, (void*)0);
+
```
Comment (drossetti) [inline tests/copylat.cpp]: please print a warning to the user re to_mapping not implying visibility to pre-launched GPU work, as stated in the GRDMA whitepaper

**Comment by pakmarkthub** [comment]
Fixed the code.

@drossetti Please re-review.

File: README.md [hunk 7]
```diff
@@ -140,32 +140,33 @@ page offset: 0
 user-space pointer: 0x7f6ae5000000
 
 gdr_copy_to_mapping num iters for each size: 10000
+WARNING: gdr_copy_to_mapping does not guarantee the visibility of data to pre-launched CUDA kernels.
```
Comment (drossetti) [inline README.md]: can you add this warning to the inline doc of gdr_copy_to_mapping in core.h ?

**Comment by pakmarkthub** [comment]
Done.

@drossetti Please take a look.

File: include/gdrapi.h [hunk 8]
```diff
@@ -122,7 +122,12 @@ int gdr_unmap(gdr_t g, gdr_mh_t handle, void *va, size_t size);
 
 // map_d_ptr is the user-space virtual address belonging to a mapping of a device memory buffer,
 // i.e. one returned by gdr_map()
+//
+// WARNING: The visibility of data to pre-launched kernels might be
```
Comment (drossetti) [inline include/gdrapi.h]: I would say "both integrity and ordering of data as observed by pre-launched GPU work is not guaranteed by this API. For more..."

File: tests/copylat.cpp [hunk 9]
```diff
@@ -213,16 +213,14 @@ int main(int argc, char *argv[])
         // gdr_copy_to_mapping benchmark
         cout << endl;
         cout << "gdr_copy_to_mapping num iters for each size: " << num_write_iters << endl;
-        cout << "WARNING: gdr_copy_to_mapping does not guarantee the visibility of data to pre-launched CUDA kernels." << endl;
+        cout << "WARNING: Measuring the latency of command-issuing latency. Data might not be visible to the GPU yet." << endl;
```
Comment (drossetti) [inline tests/copylat.cpp]: How about "Measuring the issue overhead as observed by the CPU. Data ..." ?

Comment (drossetti) [reply tests/copylat.cpp]: " Data might not be ordered all the way to the GPU internal visibility."

**Comment by pakmarkthub** [comment]
Fixed.

@drossetti Please re-review.

**Comment by drossetti** [comment]
@pakmarkthub if you are finished with this, feel free to go ahead and merge it

**Pull request merged**